### PR TITLE
feat(kmenu): add new kongponent

### PR DESCRIPTION
**Pull request by aanchalm01**
### Summary
Create new `KMenu` kongponent.
#### Changes made:
Add new component `KMenu`, and child component `KMenuItem`
Supports use of `items` prop or `body` slot which can instantiate `KMenuItem`
Add docs
Add tests

### PR Checklist
- [x] Does not introduce dependencies
- [x] **Functional:** all changes do not break existing APIs and if so, bump major version.
- [x] **Tests pass:** check the output of yarn test packages/<Kongponent>
- [x] **Naming:** the files and the method and prop variables use the same naming conventions as other Kongponents
- [x] **Framework style:** abides by the essential rules in Vue's style guide
- [x] **Cleanliness:** does not have formatting issues, unused code (e.g., console.logs), or leftover comments
- [x] **Docs:** includes a technically accurate README, uses JSDOC where appropriate

### CODE REVIEW

File: packages/KMenu/KMenuItem.vue [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    transform: translateX(-57%);
+    bottom: 0;
+    width: 88%;
+    height: 1px;
+    background-color: var(--grey-300);
+  }
+}
+
+.k-menu ul:not(:last-child):after {
+    content: " ";
```
Comment (chrisvdp) [inline packages/KMenu/KMenuItem.vue:118]: **nitpick:** ‏indentation

File: packages/KMenu/KMenuItem.vue [hunk 2]
```diff
@@ -0,0 +1,156 @@
+<template>
+  <div
+    :data-testid="item ? `${item.title.replace(' ', '-')}-menu-item` : 'menu-item'"
+    :test-mode="testMode"
+    :class="[isOpen && expandable ? 'title-dark' : '', {'expand-item' : expandable}]"
+    class="k-menu-item">
+    <div
+      class="menu-title"
+      @click="isOpen = !isOpen">
```
Comment (chrisvdp) [inline packages/KMenu/KMenuItem.vue:9]: **issue:** ‏we should not have click events on non clickable elements. This should be an <a> or <button>

Comment (chrisvdp) [reply packages/KMenu/KMenuItem.vue:9]: **issue:** ‏any click event should also have a keyboard equivalent

Comment (chrisvdp) [reply packages/KMenu/KMenuItem.vue:9]: can we add aria attributes for this

aria-controls
aria-expanded

File: packages/KMenu/KMenuItem.vue [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+        class="span-icon-container">
+        <KIcon
+          v-if="expandable"
+          :icon="isOpen ? 'chevronUp' : 'chevronDown'"
+          color="var(--grey-400)"
+          size="16"
+        />
+      </span>
+    </div>
+    <div
```
Comment (chrisvdp) [inline packages/KMenu/KMenuItem.vue:27]: can we add aria attributes for this

aria-labelledby

**Review by chrisvdp — CHANGES_REQUESTED** [review]
This seems like an accordion to me. Is there a reason we went with the name menu?

File: packages/KMenu/KMenuItem.vue (hunk 1 above)
Comment (aanchalm01) [reply packages/KMenu/KMenuItem.vue:118]: Fixed!

File: packages/KMenu/KMenuItem.vue (hunk 2 above)
Comment (aanchalm01) [reply packages/KMenu/KMenuItem.vue:9]: So this click event needs to be removed ?

Comment (chrisvdp) [reply packages/KMenu/KMenuItem.vue:9]: we can have the click event, but the element should be changed to a button

**Comment by aanchalm01** [comment]
@chrisvdp: The name Menu is selected because it's going to be used from inside a table for filtering/sort purposes. Created it like an accordion because that was the requirement.

File: packages/KMenu/KMenuItem.vue (hunk 2 above)
Comment (chrisvdp) [reply packages/KMenu/KMenuItem.vue:9]: for the keyboard event see https://github.com/maranran/eslint-plugin-vue-a11y/blob/master/docs/rules/click-events-have-key-events.md

Comment (aanchalm01) [reply packages/KMenu/KMenuItem.vue:9]: Okay! I see. I will add button and keyboard equivalent and aria attributes.

File: packages/KMenu/KMenuItem.vue [hunk 4]
```diff
@@ -0,0 +1,174 @@
+<template>
+  <div
+    :data-testid="item ? `${item.title.replace(' ', '-')}-menu-item` : 'menu-item'"
+    :test-mode="testMode"
+    :class="[isOpen && expandable ? 'title-dark' : '', {'expand-item' : expandable}]"
+    aria-labelledby="menu-item"
+    class="k-menu-item">
+    <button
```
Comment (chrisvdp) [inline packages/KMenu/KMenuItem.vue:8]: is there a reason not to us KButton here?

File: packages/KMenu/KMenuItem.vue [hunk 5]
```diff
@@ -0,0 +1,174 @@
+<template>
+  <div
+    :data-testid="item ? `${item.title.replace(' ', '-')}-menu-item` : 'menu-item'"
+    :test-mode="testMode"
+    :class="[isOpen && expandable ? 'title-dark' : '', {'expand-item' : expandable}]"
+    aria-labelledby="menu-item"
```
Comment (chrisvdp) [inline packages/KMenu/KMenuItem.vue:6]: I can't tell what this is associated with. It should be the id of an element.

Comment (chrisvdp) [reply packages/KMenu/KMenuItem.vue:6]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_aria-labelledby_attribute

**Review by chrisvdp — DISMISSED** [review]

File: packages/KMenu/KMenuItem.vue (hunk 4 above)
Comment (aanchalm01) [reply packages/KMenu/KMenuItem.vue:8]: The button here does not have any styles associated with what we have in KButton, therefore used general button element. Does that make sense ?

Comment (kaiarrowood) [reply packages/KMenu/KMenuItem.vue:8]: Would `non-visual-button` class help here?

image

**Comment by kaiarrowood** [comment]
> @chrisvdp: The name Menu is selected because it's going to be used from inside a table for filtering/sort purposes. Created it like an accordion because that was the requirement.

It's also not an accordion because we allow multiple entries to be expanded simultaneously.

File: packages/KMenu/KMenuItem.vue (hunk 4 above)
Comment (aanchalm01) [reply packages/KMenu/KMenuItem.vue:8]: I tried using `.non-visual-button` but it has `cursor: pointer !important;` and if the menu item is not expandable I need the cursor not to be a pointer. I will see if I can change the specificity and make it non-cursor but also: is it okay to have !important for a specific styling in KMenuItem.vue file ?

**Comment by chrisvdp** [comment]
> > @chrisvdp: The name Menu is selected because it's going to be used from inside a table for filtering/sort purposes. Created it like an accordion because that was the requirement.
> 
> It's also not an accordion because we allow multiple entries to be expanded simultaneously.

That's not always a constraint of an accordion, it's just an implementation detail. Bootstrap for example has an option to allow multiples to be open. https://getbootstrap.com/docs/5.0/components/accordion/#always-open

I think we should avoid thinking about a single use case only in a design system and think about potential reuse. We can use this component as a menu, but that doesn't mean it's the only use case.

File: packages/KMenu/KMenuItem.vue (hunk 4 above)
Comment (kaiarrowood) [reply packages/KMenu/KMenuItem.vue:8]: I think in this scenario using `!important` will be ok provided it is on a custom class we don't see outside of KMenu (or inside a scoped SCSS block).

Comment (chrisvdp) [reply packages/KMenu/KMenuItem.vue:8]: I don't think you need to render a button in the case where it is not expandable.

Comment (aanchalm01) [reply packages/KMenu/KMenuItem.vue:8]: There is a named slot inside KButton that I am using to display the item title, if I use v-if for the button, I won't be able to display the title using same named slot.

File: packages/KMenu/KMenuItem.vue (hunk 3 above)
Comment (kaiarrowood) [reply packages/KMenu/KMenuItem.vue:33]: Done

File: packages/KMenu/KMenuItem.vue (hunk 4 above)
Comment (kaiarrowood) [reply packages/KMenu/KMenuItem.vue]: All menu items are considered clickable. Fire 'clicked' event if non-expando, fire 'toggled' event if expando.

File: packages/KMenu/KMenuItem.vue (hunk 5 above)
Comment (kaiarrowood) [reply packages/KMenu/KMenuItem.vue]: Fixed.

**Review by chrisvdp — APPROVED** [review]

**Pull request merged**